File: neetcode150/medium/238-Product-of-Array-Except-Self.py
#!/usr/bin/python3
class Solution:
  # Prefix and Postfix solution O(n)
  def productExceptSelf(self, nums: list) -> list:
    n = len(nums)
    pre = [0] * n
    post = [0] * n
        
    pre[0] = 1
    for i in range(1, n):
      pre[i] = nums[i - 1] * pre[i - 1]
        
    post[n - 1] = 1
    for i in range(n-2, -1, -1):
      post[i] = nums[i + 1] * post[i + 1]
          
    s = [0] * n
    for i in range(n):
      s[i] = pre[i] * post[i]
        
    return s
  
  # Optimal Solution with O(n) runtime with O(1) space (output array doesn't count)
  def productExceptSelf(self, nums: list) -> list:
    res = [1] * len(nums)
        
    pre = 1
    for i in range(len(nums)):
      res[i] = pre 
      pre *= nums[i]
        
    post = 1
    for i in range(len(nums)-1, -1, -1):
      res[i] *= post 
      post *= nums[i]
            
    return res

File: neetcode150/medium/cpp/238-Product-of-Array-Except-Self.cpp
#include <vector>

using namespace std;

// O(1) extra space solution (The output array does not count as extra space for space complexity analysis).
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int N = nums.size();
        vector<int> out(N);
        
        int t = 1;
        for (int i = 0; i < N; i++){
            out[i] = t;
            t *= nums[i];
        }

        t = 1;
        for (int i = N-1; i >= 0; i--){
            out[i] *= t;
            t *= nums[i];
        }
        return out;
    }
};